File: HW4/Problem2.m
%% Problem 2
% $$\left[\matrix{ -M_A\omega^2+\gamma_2+\gamma_1 & -\gamma_1 & -\gamma_2 e^{-ika}\cr 
% -\gamma_1 & -M_B\omega^2+2\gamma_1 & -\gamma_1\cr -\gamma_2 e^{ika} & -\gamma_1 
% & -M_A\omega^2+\gamma_1+\gamma_2} \right] \left[\matrix { u_{A1}\cr u_B\cr u_{A2}}\right]= 
% 0$$

clc
clear
close all
%% 
% $$M_A/M_B = 1$$
% 
% $$\gamma_1/\gamma_2 = 1$$

gamma1 = 1;
gamma2 = 1;
MA = 1;
MB = 1;
a = 1;

k = linspace(0, pi/a, 50); % 50 k-values
w = zeros(length(k), 3);   % Preallocate 50x3 matrix for roots

% Constant coefficients
A = -MA^2 * MB;
B = (2 * MA^2 * gamma1) + (2 * MA * MB * gamma1) * (2 * MA * MB * gamma2);
C = (-2 * MA * gamma1^2) - (4 * MA * gamma1 * gamma2) - (MB * gamma1^2) - (2 * MB * gamma1 * gamma2);

for i = 1:length(k)
    D = (-2 * gamma1^2 * gamma2 * cos(k * a)) + (2 * gamma1^2 * gamma2);

    coeffs = [A, B, C, D];
    roots_i = roots(coeffs);

    % Filter real, non-negative roots and take sqrt to get 
    omega_sq = real(roots_i(imag(roots_i) == 0 & real(roots_i) >= 0));
    omega_sq = sort(omega_sq); % Optional: sort low to high
    n_roots = length(omega_sq);
    w(i, 1:n_roots) = sqrt(omega_sq);
end

% Plot the three branches
figure;
plot(k, w, 'LineWidth', 1.5);
xlabel('k'); ylabel('\omega(k)');
title('Phonon Dispersion for 3-atom Basis (A–B–A)');
legend('\omega_1', '\omega_2', '\omega_3');
grid on;
%% 
% $$M_A/M_B = 10$$
% 
% $$\gamma_1/\gamma_2 = 1$$

gamma1 = 1;
gamma2 = 1;
MA = 10;
MB = 1;

w = zeros(length(k), 3);   % Preallocate 50x3 matrix for roots

% Constant coefficients
A = -MA^2 * MB;
B = (2 * MA^2 * gamma1) + (2 * MA * MB * gamma1) * (2 * MA * MB * gamma2);
C = (-2 * MA * gamma1^2) - (4 * MA * gamma1 * gamma2) - (MB * gamma1^2) - (2 * MB * gamma1 * gamma2);

for i = 1:length(k)
    D = (-2 * gamma1^2 * gamma2 * cos(k * a)) + (2 * gamma1^2 * gamma2);

    coeffs = [A, B, C, D];
    roots_i = roots(coeffs);

    % Filter real, non-negative roots and take sqrt to get 
    omega_sq = real(roots_i(imag(roots_i) == 0 & real(roots_i) >= 0));
    omega_sq = sort(omega_sq); % Optional: sort low to high
    n_roots = length(omega_sq);
    w(i, 1:n_roots) = sqrt(omega_sq);
end

% Plot the three branches
figure;
plot(k, w, 'LineWidth', 1.5);
xlabel('k'); ylabel('\omega(k)');
title('Phonon Dispersion for 3-atom Basis (A–B–A)');
legend('\omega_1', '\omega_2', '\omega_3');
grid on;
%% 
% $$M_A/M_B = 1$$
% 
% $$\gamma_1/\gamma_2 = 10$$

gamma1 = 10;
gamma2 = 1;
MA = 1;
MB = 1;

w = zeros(length(k), 3);   % Preallocate 50x3 matrix for roots

% Constant coefficients
A = -MA^2 * MB;
B = (2 * MA^2 * gamma1) + (2 * MA * MB * gamma1) * (2 * MA * MB * gamma2);
C = (-2 * MA * gamma1^2) - (4 * MA * gamma1 * gamma2) - (MB * gamma1^2) - (2 * MB * gamma1 * gamma2);

for i = 1:length(k)
    D = (-2 * gamma1^2 * gamma2 * cos(k * a)) + (2 * gamma1^2 * gamma2);

    coeffs = [A, B, C, D];
    roots_i = roots(coeffs);

    % Filter real, non-negative roots and take sqrt to get 
    omega_sq = real(roots_i(imag(roots_i) == 0 & real(roots_i) >= 0));
    omega_sq = sort(omega_sq); % Optional: sort low to high
    n_roots = length(omega_sq);
    w(i, 1:n_roots) = sqrt(omega_sq);
end

% Plot the three branches
figure;
plot(k, w, 'LineWidth', 1.5);
xlabel('k'); ylabel('\omega(k)');
title('Phonon Dispersion for 3-atom Basis (A–B–A)');
legend('\omega_1', '\omega_2', '\omega_3');
grid on;
%% 
% $$M_A/M_B = 10$$
% 
% $$\gamma_1/\gamma_2 = 10$$

gamma1 = 10;
gamma2 = 1;
MA = 10;
MB = 1;

w = zeros(length(k), 3);   % Preallocate 50x3 matrix for roots

% Constant coefficients
A = -MA^2 * MB;
B = (2 * MA^2 * gamma1) + (2 * MA * MB * gamma1) * (2 * MA * MB * gamma2);
C = (-2 * MA * gamma1^2) - (4 * MA * gamma1 * gamma2) - (MB * gamma1^2) - (2 * MB * gamma1 * gamma2);

for i = 1:length(k)
    D = (-2 * gamma1^2 * gamma2 * cos(k * a)) + (2 * gamma1^2 * gamma2);

    coeffs = [A, B, C, D];
    roots_i = roots(coeffs);

    % Filter real, non-negative roots and take sqrt to get 
    omega_sq = real(roots_i(imag(roots_i) == 0 & real(roots_i) >= 0));
    omega_sq = sort(omega_sq); % Optional: sort low to high
    n_roots = length(omega_sq);
    w(i, 1:n_roots) = sqrt(omega_sq);
end

% Plot the three branches
figure;
plot(k, w, 'LineWidth', 1.5);
xlabel('k'); ylabel('\omega(k)');
title('Phonon Dispersion for 3-atom Basis (A–B–A)');
legend('\omega_1', '\omega_2', '\omega_3');
grid on;

File: HW4/Problem2.py
import sympy as sp
import numpy as np
import matplotlib.pyplot as plt

# Define symbols
MA, MB = sp.symbols('M_A M_B', real=True)
gamma1, gamma2 = sp.symbols('gamma_1 gamma_2', real=True)
omega, a, k = sp.symbols('omega a k', real=True)
x = sp.symbols('x', real=True)  # x = omega^2
i = sp.I

# Parameter values
params = {
    gamma1: 1.0,
    gamma2: 1.0,
    MA: 10.0,
    MB: 1.0,
    a: 1.0
}
a_val = params[a]

# Define symbolic matrix
M = sp.Matrix([
    [-MA * omega**2 + gamma2 + gamma1, -gamma1, -gamma2 * sp.exp(-i * k * a)],
    [-gamma1, -MB * omega**2 + 2 * gamma1, -gamma1],
    [-gamma2 * sp.exp(i * k * a), -gamma1, -MA * omega**2 + gamma1 + gamma2]
])

# Compute and simplify determinant
detM = sp.simplify(M.det()).subs(omega**2, x)
detM_sub = detM.subs(params)

# Extract coefficients as functions of k
coeffs_sym = sp.Poly(detM_sub, x).all_coeffs()
coeffs_sym = [sp.simplify(c) for c in coeffs_sym]
coeff_funcs = [sp.lambdify(k, c, modules='numpy') for c in coeffs_sym]

# Solve cubic in x = ω² for each k
k_vals = np.linspace(0, np.pi / a_val, 200)
omega_branches = [[], [], []]

for k_val in k_vals:
    coeff_vals = [f(k_val) for f in coeff_funcs]
    roots = np.roots(coeff_vals)
    real_omegas = [np.sqrt(r.real) for r in roots if np.isreal(r) and r.real >= 0]
    real_omegas.sort()
    for i in range(3):
        omega_branches[i].append(real_omegas[i] if i < len(real_omegas) else np.nan)

omega_branches = np.array(omega_branches)

# Plotting
plt.figure(figsize=(8, 5))
for i in range(3):
    label = f'Optical Branch {i}' if i > 0 else 'Acoustic'
    plt.plot(k_vals, omega_branches[i], label=label)

plt.xlabel(r'$k\ (\pi/a)$')
plt.ylabel(r'$\omega$ (rad/s)')
plt.title(r'Phonon Dispersion: $\omega(k)$ for 3-atom unit cell')
plt.grid(True)
plt.legend()

# Add parameter text box
param_text = (f"$\gamma_1/\gamma_2 = {params[gamma1]/params[gamma2]:.1f}$\n"
              f"$M_A = {params[MA]}$\n"
              f"$M_B = {params[MB]}$")
plt.text(0.05, 0.95, param_text, transform=plt.gca().transAxes,
         fontsize=10, verticalalignment='top',
         bbox=dict(boxstyle='round', facecolor='white', alpha=0.8))

plt.tight_layout()
plt.savefig("phonon_dispersion.pdf", format='pdf', bbox_inches='tight')
plt.show()
